lib/syslog/pool/pool.go

```diff
@@ -0,0 +1,79 @@
+package pool
+
+import (
+	"io"
+)
+
+// A LimitedConnPool is a connection pool, with the property that
+// the total number of live connections is limited to the size
+// parameter passed to New.
+type LimitedConnPool struct {
+	conns  chan io.WriteCloser // Available connections in the pool
+	live   chan struct{}       // Keep a count of living connections (in our hands, or the client)
+	signal chan struct{}       // Used to wake up the connection producer
+	err    chan error          // Send dial errors back to the client
+}
+
+// New returns a new LimitedConnPool.
+func New(dial func() (io.WriteCloser, error), size int) (*LimitedConnPool, error) {
+	// Tentative first try - if this doesn't work, we assume it never will
+	// and fail to initialize.
+	w, err := dial()
+	if err != nil {
+		return nil, err
+	}
+
+	p := LimitedConnPool{
+		conns:  make(chan io.WriteCloser, size),
+		live:   make(chan struct{}, size),
+		signal: make(chan struct{}),
+		err:    make(chan error),
+	}
+
+	p.conns <- w
+	p.live <- struct{}{}
+
+	// keep p.conns populated
+	go func() {
+		for range p.signal {
+			for len(p.live) < size {
+				w, err := dial()
+				if err != nil {
+					p.err <- err
```

Comment:
Should we make this a select with a default so it's non-blocking, otherwise if nothing is reading on the errors channel it blocks the normal execution of the connection pool.

Reply:
I agree, in fact was just writing this select statement when I saw your comment :D

Reply:
Done, and allowed a buffer of errors up to the pool size.